security/firewall_restriction.rst

```diff
@@ -4,19 +4,77 @@
 How to Restrict Firewalls to a Request
 ======================================
 
-When using the Security component, you can create firewalls that match certain request options.
-In most cases, matching against the URL is sufficient, but in special cases, you can further
-restrict the initialization of a firewall against other options of the request.
+When using the Security component, firewalls will decide whether they handle a request based on the
+result of a request matcher: the first firewall matching the request will handle it.
+
+The last firewall can be configured without any matcher to handle every incoming request.
+
+Restricting by Service
+----------------------
+
+You can configure any service implementing :class:`Symfony\\Component\\HttpFoundation\\RequestMatcherInterface`
+as ``request_matcher``.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/security.yaml
+
+        # ...
+        security:
+            firewalls:
+                secured_area:
+                    request_matcher: app.firewall.secured_area.request_matcher
+                    # ...
+
+    .. code-block:: xml
+
+        <!-- config/packages/security.xml -->
+        <?xml version="1.0" encoding="UTF-8"?>
+        <srv:container xmlns="http://symfony.com/schema/dic/security"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:srv="http://symfony.com/schema/dic/services"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <config>
+                <!-- ... -->
+                <firewall name="secured_area" request-matcher="app.firewall.secured_area.request_matcher">
+                    <!-- ... -->
+                </firewall>
+            </config>
+        </srv:container>
+
+    .. code-block:: php
+
+        // config/packages/security.php
+
+        // ...
+        $container->loadFromExtension('security', [
+            'firewalls' => [
+                'secured_area' => [
+                    'request_matcher' => 'app.firewall.secured_area.request_matcher',
+                    // ...
+                ],
+            ],
+        ]);
+
+However in most cases you don’t need to create these matchers yourself as Symfony can do it for you based
```

Comment:
```suggestion
However in most cases you don't need to create these matchers yourself as Symfony can do it for you based
```

I don't think we use these nice typographically correct characters in the documentation (as most of us write the docs in their favorite IDE, so you'll always get the developer-single-quotes).